man/vdlimit.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.41.2.
.TH VDLIMIT "8" "May 2013" "vdlimit  -- manages disk limits" "System Administration"
.SH NAME
vdlimit \- manages disk limits
.SH SYNOPSIS
.B vdlimit
\fI--xid <xid> \fR[\fI--flags <flags>\fR] \fI(--set <limit>=<value>|--remove) <mount point>\fR
.SH DESCRIPTION
.TP
\fB\-\-set\fR|\-s <limit>=<value>
\&...  set <limit> to <value>, where limit is
one of: space_used, space_total, inodes_used,
inodes_total, reserved
.TP
\fB\-\-remove\fR|\-d
\&...  removes the disk limit for <xid> from <mount point>
.PP
Please report bugs to vserver@list.linux\-vserver.org
.PP
This program is part of util\-vserver
.SH COPYRIGHT
Copyright \(co 2005 Enrico Scholz
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.


NEWS
version 0.30.215
================

	- added support for using libnss instead of beecrypt for
	vhashify's hash calculation.  libnss has bad SHA1 performance on
	i386, but is a) maintained and b) gives better performance with
	all other hashes and on x86_64.

	- added tunctl, a small program to setup tun/tap interfaces, with
	support in the configuration.

	- vattribute and nattribute now support a --get operation, which
	parses the flags and capabilities into a human-readable form.

	- vspace supports all the spaces currently available.

	- vmemctrl controls the OOM bias of context.

	- dynamic contexts are unbroken.

	- hopefully all rpm-fake races have been fixed.

	- flock has been replaced by lockf, which works on NFS.

	- the clone build method now supports exclude lists.

	- vmount handles mounting into a guest.

	- various fixes for the Gentoo-scripts, thanks to Benedikt Böhm.

	- support for yum's persistdir has been added.

	- per-guest device maps are now supported.

	- vserver ... build now sets the barrier on the parent directory.

	- fsck is now supported for the guest's fstab.

	- symlinks are now handled on all levels in namespace cleanup.

	- the debootstrap build method now supports debootstrap 1.0.8.

	- a documentation update, thanks to Micah Anderson.

	- use numeric uids/gids for the template build method.

	- add support for the alpine and f8 distributions.
	
version 0.30.214
================
	- patches for yum-3.2.

	- various fixes to allow for *SuSE guests.

	- added Fedora 7.

	- replace file with h2ext for build -m template.

	- add exclude list support for vclone.

	- include debootstrap scripts for a lot of distributions.

	- fixes to allow the Gentoo initscripts to be used with baselayout 2.

	- remove disk limit caches on vserver ... delete.

	- added support for the new APIs in 2.2 and 2.3, in particular the
	netv2 API, allowing native IPv6 support.

	- added support for a separate filesystem tag.

version 0.30.213
================
	- added support for the device mapping feature present in development
	kernels.

	- further improved Gentoo support, again thanks to Benedikt Böhm.

	- support persistent contexts.

	- add another initscript that creates directories, configures vshelper
	as well as kills all contexts.

	- new vserver-stat implementation, which uses the accounting APIs
	available in Linux-VServer 2.2.

	- added support for setting sysctl values in the guest configuration.

	- added a clone build method, which simply links already unified files.

	- implement dynamic contexts in userspace, generating a static one
	during vserver ... build.

	- fix the endless loop in vlogin.

	- fix Ubuntu builds.

	- improved support for Debian Etch, thanks to Benedikt Böhm.

	- added CentOS 5.

	- use the state change helper to try to cleanup after us if the guest
	exited on its own.

version 0.30.212
================
	- support for the Linux-VServer 2.1 API has been implemented.

	- the scheduler configuration has been split up into multiple files.

	- namespace cleanup is enabled by default. This can be disabled by
	touch ${sysconfdir}/vservers/.defaults/nonamespace-cleanup

	- all unset ulimits are now reset to infinity, so limits are no longer
	inherited from the host.

	- much improved Gentoo support, thanks to Benedikt Böhm.

version 0.30.211
================

	- novlandev is now the default. To reenable the old behaviour of
	creating VLAN interfaces where used, touch
	${sysconfdir}/vservers/.defaults/interfaces/vlandev.

	- vshelper now works correctly with the kernel legacy support turned
	off.

	- vlogin provides a terminal proxy which allocates a new psuedo-tty
	on the inside when you enter a guest.

	- and lots of other features/fixes, see ChangeLog.


version 0.30.210
================

	- LEGACY BREAKAGE: by default, only recent kernel APIs will be
	supported so you have to make sure to use at least the vs2.0.1
	kernel patch. With older kernels, you have to call 'configure'
	with '--enable-apis=NOLEGACY'. Some programs will not work anymore
	with the default settings; these ones are:

	     + capchroot
	     * chcontext-context
	     * reducecap

	- dlimit support was integrated into the standard vserver
	operations. Thanks to Daniel Hokka Zakrisson for his patch.

	- recent kernel API is now used for setting up the network
	contexts. So it should be now possible to use a kernel compiled
	without CONFIG_VSERVER_LEGACYNET. Thanks to Daniel Hokka Zakrisson
	for his patch.  .


version 0.30.209
================

	- large revamping...
	
	
version 0.23.6
==============

	- fixed '--level' option on 'vserver ... chkconfig'


src/vdlimit.c
// $Id$    --*- c -*--

// Copyright (C) 2005 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "util.h"
#include <lib/internal.h>

#include <vserver.h>

#include <getopt.h>
#include <libgen.h>
#include <errno.h>
#include <signal.h>
#include <sched.h>

#define ENSC_WRAPPERS_PREFIX	"vdlimit: "
#define ENSC_WRAPPERS_UNISTD	1
#define ENSC_WRAPPERS_VSERVER	1
#include <wrappers.h>

#define CMD_HELP		0x1000
#define CMD_VERSION		0x1001

int		wrapper_exit_code  =  1;

struct option const
CMDLINE_OPTIONS[] = {
  { "help",       no_argument,       0, CMD_HELP },
  { "version",    no_argument,       0, CMD_VERSION },
  { "xid",        required_argument, 0, 'x' },
  { "set",        required_argument, 0, 's' },
  { "remove",     no_argument,       0, 'd' },
  { "flags",      required_argument, 0, 'f' },
  {0,0,0,0}
};

static void
showHelp(int fd, char const *cmd)
{
  WRITE_MSG(fd, "Usage: ");
  WRITE_STR(fd, cmd);
  WRITE_MSG(fd,
	    " --xid <xid> [--flags <flags>] (--set <limit>=<value>|--remove) <mount point>\n"
	    "\n"
	    "    --set|-s <limit>=<value>  ...  set <limit> to <value>, where limit is \n"
	    "                           one of: space_used, space_total, inodes_used,\n"
	    "                           inodes_total, reserved\n"
	    "    --remove|-d       ...  removes the disk limit for <xid> from <mount point>\n"
	    "\n"
	    "Please report bugs to " PACKAGE_BUGREPORT "\n");

  exit(0);
}

static void
showVersion()
{
  WRITE_MSG(1,
	    "vdlimit " VERSION " -- manages disk limits\n"
	    "This program is part of " PACKAGE_STRING "\n\n"
	    "Copyright (C) 2005 Enrico Scholz\n"
	    VERSION_COPYRIGHT_DISCLAIMER);
  exit(0);
}

static void
setDlimit(char const *filename, xid_t xid, uint32_t flags, struct vc_ctx_dlimit const *limit)
{
  bool		was_added = false;

  if (vc_get_dlimit(filename, xid, flags, 0) == -1) {
    if (vc_add_dlimit(filename, xid, flags) == -1) {
      perror(ENSC_WRAPPERS_PREFIX "vc_add_dlimit()");
      exit(wrapper_exit_code);
    }

    was_added = true;
  }

  if (vc_set_dlimit(filename, xid, flags, limit) == -1) {
    perror(ENSC_WRAPPERS_PREFIX "vc_set_dlimit()");

    if (was_added &&
	vc_rem_dlimit(filename, xid, flags)==-1)
      perror(ENSC_WRAPPERS_PREFIX "vc_rem_dlimit()");

    exit(wrapper_exit_code);
  }
}

static void
remDlimit(char const *filename, xid_t xid, uint32_t flags)
{
  if (vc_rem_dlimit(filename, xid, flags) == -1) {
    perror(ENSC_WRAPPERS_PREFIX "vc_rem_dlimit()");
    exit(wrapper_exit_code);
  }
}

static void
writeInt(int fd, char const *prefix, unsigned int val)
{
  char		buf[sizeof(val)*3 + 2];
  size_t	len = utilvserver_fmt_uint(buf, val);

  if (prefix)
    WRITE_STR(fd, prefix);
  if (val == VC_CDLIM_INFINITY)
    WRITE_STR(fd, "-1");
  else
    Vwrite(fd, buf, len);
}

static void
printDlimit(char const *filename, xid_t xid, uint32_t flags, bool formatted)
{
  struct vc_ctx_dlimit		limit;
  
  if (vc_get_dlimit(filename, xid, flags, &limit) == -1) {
    perror(ENSC_WRAPPERS_PREFIX "vc_get_dlimit()");
    exit(wrapper_exit_code);
  }

  if (formatted) {
    writeInt (1, 0, xid);
    WRITE_MSG(1, " ");
    WRITE_STR(1, filename);
    writeInt (1, "\nspace_used=",   limit.space_used);
    writeInt (1, "\nspace_total=",  limit.space_total);
    writeInt (1, "\ninodes_used=",  limit.inodes_used);
    writeInt (1, "\ninodes_total=", limit.inodes_total);
    writeInt (1, "\nreserved=",     limit.reserved);
    WRITE_MSG(1, "\n");
  }
  else {
    writeInt (1, 0,   xid);
    writeInt (1, " ", limit.space_used);
    writeInt (1, " ", limit.space_total);
    writeInt (1, " ", limit.inodes_used);
    writeInt (1, " ", limit.inodes_total);
    writeInt (1, " ", limit.reserved);
    WRITE_MSG(1, " ");
    WRITE_STR(1, filename);
    WRITE_MSG(1, "\n");
  }
}


static bool
setDLimitField(struct vc_ctx_dlimit *dst, char const *opt)
{
  uint_least32_t	*ptr;
  char const * const	orig_opt = opt;

#define GET_VAL_PTR(CMP, VAL)						\
  (strncmp(opt, CMP "=", sizeof(CMP))==0) ?				\
  (opt+=sizeof(CMP), &VAL) : 0
  
  if      ((ptr=GET_VAL_PTR("space_used",   dst->space_used))!=0)   {}
  else if ((ptr=GET_VAL_PTR("space_total",  dst->space_total))!=0)  {}
  else if ((ptr=GET_VAL_PTR("inodes_used",  dst->inodes_used))!=0)  {}
  else if ((ptr=GET_VAL_PTR("inodes_total", dst->inodes_total))!=0) {}
  else if ((ptr=GET_VAL_PTR("reserved",     dst->reserved))!=0)     {}
  else      ptr=0;

#undef  GET_VAL_PTR  

  if (ptr!=0 && *ptr==VC_CDLIM_KEEP) {
    char	*endptr;
    long	val = strtol(opt, &endptr, 0);

    if (*opt==0 || *endptr!='\0') {
      WRITE_MSG(2, ENSC_WRAPPERS_PREFIX "can not parse number in '");
      WRITE_STR(2, orig_opt);
      WRITE_MSG(2, "'\n");
      return false;
    }

    *ptr = val;
  }
  else if (ptr!=0) {
    WRITE_MSG(2, ENSC_WRAPPERS_PREFIX "value already set in '");
    WRITE_STR(2, orig_opt);
    WRITE_MSG(2, "'\n");
    return false;
  }
  else {
    WRITE_MSG(2, ENSC_WRAPPERS_PREFIX "unknown limit in '");
    WRITE_STR(2, orig_opt);
    WRITE_MSG(2, "'\n");
    return false;
  }

  return true;
}

bool
isHigherLimit(uint_least32_t lhs, uint_least32_t rhs)
{
  if (lhs==VC_CDLIM_KEEP || rhs==VC_CDLIM_KEEP) return false;

  return lhs > rhs;
}

int main(int argc, char *argv[])
{
  bool		do_set       = false;
  bool		do_remove    = false;
  xid_t		xid          = VC_NOCTX;
  uint32_t	flags        = 0;
  char		*endptr;
  int		sum          = 0;

  struct vc_ctx_dlimit		limit = {
    .space_used   = VC_CDLIM_KEEP,
    .space_total  = VC_CDLIM_KEEP,
    .inodes_used  = VC_CDLIM_KEEP,
    .inodes_total = VC_CDLIM_KEEP,
    .reserved     = VC_CDLIM_KEEP
  };
  
  while (1) {
    int		c = getopt_long(argc, argv, "+x:s:df:", CMDLINE_OPTIONS, 0);
    if (c==-1) break;

    switch (c) {
      case CMD_HELP	:  showHelp(1, argv[0]);
      case CMD_VERSION	:  showVersion();
      case 'x'		:  xid = Evc_xidopt2xid(optarg, true); break;
      case 's'		:
	if (!setDLimitField(&limit, optarg))
	  return EXIT_FAILURE;
	else
	  do_set = true;
	break;
      case 'd'		:  do_remove = true; break;
      case 'f'		:
	{
	  flags = strtol(optarg, &endptr, 0);
	  if ((flags == 0 && errno != 0) || *endptr != '\0') {
	    WRITE_MSG(2, "Invalid flags argument: '");
	    WRITE_STR(2, optarg);
	    WRITE_MSG(2, "'; try '--help' for more information\n");
	    return EXIT_FAILURE;
	  }
	}
	break;

      default		:
	WRITE_MSG(2, "Try '");
	WRITE_STR(2, argv[0]);
	WRITE_MSG(2, " --help' for more information.\n");
	return EXIT_FAILURE;
	break;
    }
  }

  sum = ((do_set ? 1 : 0) + (do_remove ? 1 : 0));
  
  if (sum>1)
    WRITE_MSG(2, "Can not specify multiple operations; try '--help' for more information\n");
  else if (optind==argc)
    WRITE_MSG(2, "No mount point specified; try '--help' for more information\n");
  else if (xid==VC_NOCTX)
    WRITE_MSG(2, "No xid specified; try '--help' for more information\n");
  else if (isHigherLimit(limit.space_used, limit.space_total))
    WRITE_MSG(2, "invalid parameters: 'space_used' is larger than 'space_total'\n");
  else if (isHigherLimit(limit.inodes_used, limit.inodes_total))
    WRITE_MSG(2, "invalid parameters: 'inodes_used' is larger than 'inodes_total'\n");
  else {
    for (; optind < argc; ++optind) {
      if      (do_set)     setDlimit(argv[optind], xid, flags, &limit);
      else if (do_remove)  remDlimit(argv[optind], xid, flags);
      else                 printDlimit(argv[optind], xid, flags, true);
    }

    return EXIT_SUCCESS;
  }

  return EXIT_FAILURE;
}
